### ISSUE TITLE
Failed ABI checks

### ISSUE THREAD

**Issue by kmoore134** [jrmarino/synth#25]
I'm trying to do a build of 10.3-BETA1 package and keep getting failed ABI checks whenever we start a new build. (Which of course requires rebuilding pkgng, and all other packages)

root@bellicose:/synth/world # file bin/sh
bin/sh: ELF 64-bit LSB executable, x86-64, version 1 (FreeBSD), dynamically linked, interpreter /libexec/ld-elf.so.1, for FreeBSD 10.2 (1002509), stripped
root@bellicose:/synth/world # ./bin/freebsd-version 
10.3-BETA1

How is the ABI determined? Anything else I can run to figure out why this fails?

**Comment by kmoore134** [comment]
BTW, I should have mentioned, the host system is running -CURRENT, but we are building 10.3-BETA1 packages using a 10.3-BETA1 base in /synth/world. I assume this is supported?

**Comment by jrmarino** [comment]
heh.  This is where synth differs from poudriere.
synth is using a filtered version of the host (by default).  So it looks like you're fixing 10.2 and 10.3.  However, that still shouldn't matter wrt to the ABI.

what does pkg -VV | grep ABI show?

in the meantime, I'll tell you where the ABI code is.

**Comment by jrmarino** [comment]
here is interesting: https://github.com/jrmarino/synth/blob/master/src/portscan-packages.adb#L731

It's calculating it from the uname.   What does "uname -mr" show?

**Comment by kmoore134** [comment]
root@bellicose:/synth/pkg # pkg -vv | grep ABI
ABI = "FreeBSD:11:amd64";
ALTABI = "freebsd:11:x86:64";
root@bellicose:/synth/pkg # uname -mr
11.0-CURRENTJAN2016 amd64

**Comment by jrmarino** [comment]
hmm, that uname should led to the same ABI.  the vm I have here is 11.0-CURRENT i386 and that has no issues.

**Comment by jrmarino** [comment]
what packages is it complaining about?  packages that synth built previously?

**Comment by kmoore134** [comment]
giving it another whirl now. The existing packages its checking originally came from a poudriere build, would that matter?

**Comment by jrmarino** [comment]
probably yes, give it a clean directory

**Comment by jrmarino** [comment]
if you are really curiously, use pkg query to find out ABI of any package in that poudriere direcotry

**Comment by jrmarino** [comment]
pkg  query -F <path to package> '%q' I think

**Comment by kmoore134** [comment]
# pkg query -F pcbsd-utils-qt5-1454017176.txz '%q'
freebsd:10:x86:64

**Comment by kmoore134** [comment]
interesting, is it because its x86:64 and not x86_64?

**Comment by jrmarino** [comment]
well, there you go

**Comment by jrmarino** [comment]
no, it's because it's 10 and not 11

**Comment by kmoore134** [comment]
ok, that begs the question then. Should this work? Or will all 10.x packages always fail ABI check when building on 11.x host?

**Comment by kmoore134** [comment]
I.E. our build nodes for 10.2 / 10.3 / 11-CURRENT all run -CURRENT. With synth I'm giving it a 10.3 world in 'Directory_system= /synth/world'

**Comment by jrmarino** [comment]
loaded question.
If your 11 host builds a 10 world and installs it, and a synth profile uses that 10 sysroot, you can build 10 packages on 11.

**Comment by jrmarino** [comment]
but if you use the live system, which is 11, then it's going to be limited to 11 packages.
poudriere does the same thing with jails

**Comment by kmoore134** [comment]
maybe I'm misunderstanding the docs then. Is using the Directory_system= the right way to give it a 10 sysroot?

**Comment by jrmarino** [comment]
> I.E. our build nodes for 10.2 / 10.3 / 11-CURRENT all run -CURRENT. With synth I'm giving it a 10.3 world in 'Directory_system= /synth/world'

ok, logic flaw on my part

**Comment by jrmarino** [comment]
checking uname isn't going to work

**Comment by jrmarino** [comment]
kmoore: you are doing it right, synth method of getting abi is logically flawed

**Comment by kmoore134** [comment]
ok, whew ;) Do you want to ping back when you have a fix? I'll be eager to test it

**Comment by jrmarino** [comment]
let me think about it a bit.  I'll ping you back
It will be easy to rebuild with github hash

**Comment by jrmarino** [comment]
by the way, it does use the uname in the specified sysroot, but I forgot that uname just reads the running kernel

**Comment by kmoore134** [comment]
BTW, while I've got you doing synth things ;)

It would be cool if we had a 'cleanup' sub-command which only does the umount routines. We have things run via Jenkins, and that would be handy if something crashed / didn't cleanup properly.

**Comment by jrmarino** [comment]
it already exists.
run "synth configure" (hit enter)
Any command (other than help and version) will clean up dirty mounts

**Comment by kmoore134** [comment]
Will that work via script? I.E. in automation for jenkins?

**Comment by jrmarino** [comment]
for jenkins, you won't be able to use any saved options configuration (unless you script that they are always valid, otherwise synth will stop before it starts).  However, I assume you don't override options anyway.

I don't see any command that will exit without doing anything substantial.

However, in theory, the next command that the script sends will clean it up.  The only way it would fail is if a process is still running which prevents the umount.

so in other words, if the mounts are dirty and a build command is issued, the mounts will always be clean up before anything else.  they don't need a special command.

**Comment by jrmarino** [comment]
okay Kris, I have something for you to test.

On the makefile line 24, change the hash value to "24bc284"
then run "make makefile"
then the whole clean/disinstall/install routine.
It should work as expected now.

**Comment by jrmarino** [comment]
I mean "makesum" (edited above)

**Comment by kmoore134** [comment]
Getting closer now, no ABI errors, but each package fails with the following:

========================< phase : check-sanity    >========================
make: "/xports/Mk/bsd.port.mk" line 1162: UNAME_r (11.0-CURRENTJAN2016) and OSVERSION (1002509) do not agree on major version number.
===========================================================================

This is expected, you'll need to also set UNAME_r env variable before starting a builder, i remember doing this back in the day when I was doing manual builds ;)

**Comment by jrmarino** [comment]
there's a workaround right now.  You can /etc/local/etc/synth/<profile>-environment and set it manually.

It's a freebsd-specific issue.  I wouldn't have seen it on dragonfly

**Comment by jrmarino** [comment]
I think the solution is related to the work I just did.  Synth is caching _OSVERSION and it's caching the host version.  This is the same logic flaw as uname (the difference being I tried to fix the former but failed and I missed this one)

So the same "file /bin/sh" solution for ABI can be reused to set _OSVERSION correctly.

**Comment by jrmarino** [comment]
ok, code is reading <sys/param.h> so that's what I need to do as well.

**Comment by kmoore134** [comment]
Ok, setting UNAME_r=10.3-BETA1 in /usr/local/etc/synth/<profile>-environment worked fine. 

Do you want to add something to set that environment variable automatically though? Thats what poudriere does, if you set it to the output of the 'freebsd-version' from the new world directory, it should work.

**Comment by jrmarino** [comment]
no, I'm going to fix it properly.

**Comment by kmoore134** [comment]
Ok, works for me. I'm going to keep testing with the workaround in the meantime, see if we run into any other issues.

**Comment by jrmarino** [comment]
okay.  I'll have a fix later today.  I'm working on it now.

**Comment by jrmarino** [comment]
This should be the proper fix: f676f03
The only "strange" thing is the _OSRELEASE now looks something like "11.0-SYNTH".  The suffix is "-SYNTH" regardless of what it really is.  Hopefully this won't cause any issues.  I guess we will see. 

ARCH, OSVERSION, _OSRELEASE, OPSYS were all reworked.

**Comment by kmoore134** [comment]
cool! I've got some builds going now, I'll give the fix a whirl shortly

**Comment by kmoore134** [comment]
Ok, ran some tests here. With the change applied, I'm still getting errors about UNAME_r not matching:

========================< phase : check-sanity    >========================
make: "/xports/Mk/bsd.port.mk" line 1162: UNAME_r (11.0-CURRENTJAN2016) and OSVERSION (1001000) do not agree on major version number.
===========================================================================

Additionally I notice some oddness in the configure scripts:

OSVERSION=100000
_OSRELEASE=1.0-SYNTH

Shouldn't _OSRELEASE be 10.0-SYNTH?

Adding UNAME_r manually to environment fixes problems.

**Comment by jrmarino** [comment]
those are the "default" values when something failed.  
This worked for me.  Can you post part of sys/param.h, the line that defines __FreeBSD_version ?

**Comment by kmoore134** [comment]
#define __FreeBSD_version 1001000       /* Master, propagated to newvers */

**Comment by kmoore134** [comment]
This is on a 10.1-RELEASE system base BTW

**Comment by jrmarino** [comment]
no, I mean the version in the system root.

**Comment by jrmarino** [comment]
or is that it?

**Comment by kmoore134** [comment]
yea, that is the file from /synth/world/usr/include/sys/parm.h, where /synth/world is my 10.1-RELEASE base

**Comment by jrmarino** [comment]
can you find a synth log and and paste the contents of the make.conf according to the log?  There's nothing wrong with the input line as far as I can tell.

**Comment by kmoore134** [comment]
Here's a couple:

http://slexy.org/view/s2hX1CkrER

http://slexy.org/view/s2G9XiLpyn

I think you still need to set UNAME_r properly in the environment. Otherwise the port system will run 'uname -r' manually and see the wrong (host) version. Thats why poudriere has to manually set that variable in each jail / builder.

**Comment by jrmarino** [comment]
i promise you it won't. :)

**Comment by kmoore134** [comment]
lol ok, you are the expert, I'll let you find the way around it ;)

**Comment by jrmarino** [comment]
your log is full of errors.
The make.conf suffix is missing and everything else has "discovery error"

**Comment by jrmarino** [comment]
basically the code probably worked, but something else is in play.

**Comment by kmoore134** [comment]
Anything I can look at? What exactly is a discovery error?

**Comment by kmoore134** [comment]
also, check the second log, no discovery error there

**Comment by jrmarino** [comment]
discovery error means the executed command failed (the one in the chroot)

**Comment by kmoore134** [comment]
Ahh wait, that one may be because it was building while I manually killed synth. My bad, take a look at the lang_gcc log instead

**Comment by kmoore134** [comment]
err, no, wait on that:

root@avenger:/jenkins # synth everything
Builder mounts detected; attempting to remove them automatically ...
Dismounting successful!
Scanning entire ports tree.
Scanning existing packages.
00:06:48 => [02] Builder launched
00:06:48 => [01] Builder launched
00:06:48 => [03] Builder launched
00:06:48 => [06] Builder launched
00:06:48 => [04] Builder launched
00:06:48 => [07] Builder launched
00:06:48 => [05] Builder launched
00:06:48 => [10] Builder launched
00:06:48 => [08] Builder launched
00:06:48 => [11] Builder launched
00:06:48 => [12] Builder launched
00:06:48 => [13] Builder launched
00:06:48 => [09] Builder launched
00:06:48 => [14] Builder launched
00:06:48 => [15] Builder launched
00:06:48 => [16] Builder launched
00:07:23 => [04] 00:00:00 Failure multimedia/ffmpeg
00:07:23 => [16] 00:00:00 Failure devel/libc++
00:07:25 => [01] 00:00:00 Failure print/cups-client
00:07:27 => [08] 00:00:00 Failure lang/gnustep-base
00:07:41 => [07] 00:00:18 Success x11-toolkits/gnome-pty-helper
00:07:43 => [14] 00:00:22 Failure lang/sbcl
00:08:04 => [15] 00:00:52 Success textproc/iksemel
00:08:14 => [06] 00:01:00 Success net/glib-networking
00:08:24 => [05] 00:01:10 Success textproc/redland

**Comment by jrmarino** [comment]
```
HAVE_COMPAT_IA32_KERN=YES
CONFIGURE_MAX_CMD_LEN=262144
_SMP_CPUS=16
UID=0
ARCH=amd64
OPSYS=FreeBSD
OSVERSION=100000
_OSRELEASE=1.0-SYNTH
[ /etc/make.conf TAIL ]
```
Looks better except for OSVERSION and _OSRELEASE.
(That's a hell of a custom makefile)

**Comment by kmoore134** [comment]
ffmpeg and a few others failed almost immediately, right at the beginning. (Those are the ones with discovery errors).

And yes, we have quite a custom make.conf, lots of things specific to desktop packages ;)

**Comment by jrmarino** [comment]
any chance mounts are failing?  This is new to me.

**Comment by kmoore134** [comment]
maybe? Not seeing anything in dmesg or other obvious mount failures. It cleans up and moves onto the next builder before I can look ;)

This is with 16 concurrent builders defined, want me to drop it down?

**Comment by jrmarino** [comment]
the number shouldn't be a problem.  You can if you want but I think the same thing will happen.  I've run with 10 before without issue.

**Comment by jrmarino** [comment]
is there anything strange about what synth is mounting to? NFS or prior mounts?

At some point I can tell you what to modify to make it print out the actual error.

**Comment by kmoore134** [comment]
No NFS, just a local ZFS filesystem. It is cleaning up prior mounts first though, leme manually clean first and test again.

**Comment by jrmarino** [comment]
hold on a sec

**Comment by jrmarino** [comment]
can you try hash "a8ed62b"?
It should at least fix the OSRELEASE (I'm guessing) and maybe other stuff with luck.

**Comment by kmoore134** [comment]
ok, give me a few min, will report back

**Comment by kmoore134** [comment]
Ok, looks like it fixes the issue with OSRELEASE:

OPSYS=FreeBSD
OSVERSION=1001000
_OSRELEASE=10.1-SYNTH

I saw the new status line go by on screen, nice!

Still seeing a couple of random builders fail when it starts up, with the discovery errors though. Only happens at the initial startup, maybe when its busy spinning up all the builders at once?

**Comment by jrmarino** [comment]
if that's the case, it's an OS issue.  Maybe freebsd isn't concurrent safe?  What filesystem is being used?
maybe now is the time to halve the builders and see if anything changes.

**Comment by kmoore134** [comment]
Ok, I just kicked off another (with full 16 builders), and didn't get any of those failures this time... I'll keep an eye on it, see if I can figure out out reproducible it is. If I find something, I'll open a new bug report. 

Marking as closed though, the original issue seems to be fixed.

**Comment by jrmarino** [comment]
ok, but it would be interesting to know if this is a ZFS system.
I haven't seen anything like this on DragonFly HAMMER or DragonFly UFS

I suspect you'll see it again, randomly.

**Comment by kmoore134** [comment]
Yes, this is ZFS system. (100% ZFS on root, nothing else)

**Comment by jrmarino** [comment]
i have a fairly strong suspicion that the custom env definitions are responsible for the discovery errors.
